File: packages/r6-client/src/normalise.ts
import type { Platform, PlayerProfile, RankInfo } from "./types";
import { slugToRankInfo } from "./ranks";

export interface ProfileResponse {
  id: string;
  user_id: string;
  username: string;
  platform: string;
  level: number;
  hs: number | null; // headshot percentage
  views: number;
  last_played_at: string;
  fetched_at: string;
  updated_at: string;
  username_history: Array<{ username: string; created_at: string }>;
  leaderboard_position: number | null;
  max_rank_season: string;
  ranked_season_records: Record<string, SeasonRankRecord>;
  season_mode_records: Record<string, Record<string, SeasonModeRecord>>;
  top_operators: {
    attacker: OperatorStats[];
    defender: OperatorStats[];
  };
  bans: unknown[];
  external_bans: unknown[];
}

export interface SeasonRankRecord {
  season: string;
  rank: string;
  max_rank: string;
  rank_points: number;
  max_rank_points: number;
  rank_position?: number;
  max_rank_position?: number;
}

export interface SeasonModeRecord {
  season: string;
  mode: string;
  matches: number;
  wins: number;
  losses: number;
  abandons: number;
  wr: number;
  kills: number;
  deaths: number;
  assists: number;
  headshots: number;
  kd: number;
  kda: number;
  kpm: number;
  hs: number;
}

export interface OperatorStats {
  operator: string;
  wins: number;
  losses: number;
  kills: number;
  deaths: number;
  rounds_played: number;
}

export interface NormaliseParams {
  platform: Platform;
  response: ProfileResponse;
  currentSeason: string;
  rankedBombMode: string;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isNonNegativeNumber(value: unknown): value is number {
  return isFiniteNumber(value) && value >= 0;
}

function isNonNegativeInteger(value: unknown): value is number {
  return isNonNegativeNumber(value) && Number.isInteger(value);
}

function isPositiveInteger(value: unknown): value is number {
  return isFiniteNumber(value) && Number.isInteger(value) && value > 0;
}

function getNonNegativeInteger(record: Record<string, unknown>, key: string): number {
  const value = record[key];
  return isNonNegativeInteger(value) ? value : 0;
}

function getNonNegativeNumber(record: Record<string, unknown>, key: string): number {
  const value = record[key];
  return isNonNegativeNumber(value) ? value : 0;
}

function parseSeasonRankRecord(value: unknown): SeasonRankRecord | null {
  if (!isRecord(value)) {
    return null;
  }
  if (
    typeof value.season !== "string" ||
    value.season.length === 0 ||
    typeof value.rank !== "string" ||
    value.rank.length === 0 ||
    typeof value.max_rank !== "string" ||
    value.max_rank.length === 0 ||
    !isNonNegativeNumber(value.rank_points) ||
    !isNonNegativeNumber(value.max_rank_points)
  ) {
    return null;
  }

  return {
    season: value.season,
    rank: value.rank,
    max_rank: value.max_rank,
    rank_points: value.rank_points,
    max_rank_points: value.max_rank_points,
    ...(isPositiveInteger(value.rank_position) && { rank_position: value.rank_position }),
    ...(isPositiveInteger(value.max_rank_position) && {
      max_rank_position: value.max_rank_position,
    }),
  };
}

function parseSeasonModeRecord(value: unknown): SeasonModeRecord | null {
  if (!isRecord(value) || typeof value.season !== "string" || typeof value.mode !== "string") {
    return null;
  }

  return {
    season: value.season,
    mode: value.mode,
    matches: getNonNegativeInteger(value, "matches"),
    wins: getNonNegativeInteger(value, "wins"),
    losses: getNonNegativeInteger(value, "losses"),
    abandons: getNonNegativeInteger(value, "abandons"),
    wr: getNonNegativeNumber(value, "wr"),
    kills: getNonNegativeInteger(value, "kills"),
    deaths: getNonNegativeInteger(value, "deaths"),
    assists: getNonNegativeInteger(value, "assists"),
    headshots: getNonNegativeInteger(value, "headshots"),
    kd: getNonNegativeNumber(value, "kd"),
    kda: getNonNegativeNumber(value, "kda"),
    kpm: getNonNegativeNumber(value, "kpm"),
    hs: getNonNegativeNumber(value, "hs"),
  };
}

function parseOperatorStats(value: unknown): OperatorStats[] {
  if (!Array.isArray(value)) {
    return [];
  }

  return value.flatMap((entry): OperatorStats[] => {
    if (!isRecord(entry) || typeof entry.operator !== "string") {
      return [];
    }
    return [
      {
        operator: entry.operator,
        wins: getNonNegativeInteger(entry, "wins"),
        losses: getNonNegativeInteger(entry, "losses"),
        kills: getNonNegativeInteger(entry, "kills"),
        deaths: getNonNegativeInteger(entry, "deaths"),
        rounds_played: getNonNegativeInteger(entry, "rounds_played"),
      },
    ];
  });
}

export function parseProfileResponse(value: unknown): ProfileResponse {
  if (
    !isRecord(value) ||
    typeof value.id !== "string" ||
    value.id.trim().length === 0 ||
    typeof value.user_id !== "string" ||
    value.user_id.trim().length === 0 ||
    typeof value.username !== "string" ||
    value.username.trim().length === 0 ||
    typeof value.platform !== "string" ||
    value.platform.trim().length === 0 ||
    !isNonNegativeInteger(value.level) ||
    typeof value.max_rank_season !== "string" ||
    value.max_rank_season.trim().length === 0 ||
    !isRecord(value.ranked_season_records) ||
    !isRecord(value.season_mode_records)
  ) {
    throw new Error("Profile response is missing required fields");
  }

  const rankedSeasonRecords: Record<string, SeasonRankRecord> = {};
  for (const [season, record] of Object.entries(value.ranked_season_records)) {
    const parsed = parseSeasonRankRecord(record);
    if (parsed) {
      rankedSeasonRecords[season] = parsed;
    }
  }

  const seasonModeRecords: Record<string, Record<string, SeasonModeRecord>> = {};
  for (const [season, modes] of Object.entries(value.season_mode_records)) {
    if (!isRecord(modes)) {
      continue;
    }
    const parsedModes: Record<string, SeasonModeRecord> = {};
    for (const [mode, record] of Object.entries(modes)) {
      const parsed = parseSeasonModeRecord(record);
      if (parsed) {
        parsedModes[mode] = parsed;
      }
    }
    seasonModeRecords[season] = parsedModes;
  }

  const topOperators = isRecord(value.top_operators) ? value.top_operators : {};

  return {
    id: value.id,
    user_id: value.user_id,
    username: value.username,
    platform: value.platform,
    level: value.level,
    hs: isNonNegativeNumber(value.hs) && value.hs <= 100 ? value.hs : null,
    views: isNonNegativeNumber(value.views) ? value.views : 0,
    last_played_at: typeof value.last_played_at === "string" ? value.last_played_at : "",
    fetched_at: typeof value.fetched_at === "string" ? value.fetched_at : "",
    updated_at: typeof value.updated_at === "string" ? value.updated_at : "",
    username_history: Array.isArray(value.username_history)
      ? value.username_history.flatMap((entry) =>
          isRecord(entry) &&
          typeof entry.username === "string" &&
          typeof entry.created_at === "string"
            ? [{ username: entry.username, created_at: entry.created_at }]
            : []
        )
      : [],
    leaderboard_position: isPositiveInteger(value.leaderboard_position)
      ? value.leaderboard_position
      : null,
    max_rank_season: value.max_rank_season,
    ranked_season_records: rankedSeasonRecords,
    season_mode_records: seasonModeRecords,
    top_operators: {
      attacker: parseOperatorStats(topOperators.attacker),
      defender: parseOperatorStats(topOperators.defender),
    },
    bans: Array.isArray(value.bans) ? value.bans : [],
    external_bans: Array.isArray(value.external_bans) ? value.external_bans : [],
  };
}

function isChampionRank(slug: string | undefined | null): boolean {
  if (!slug) {
    return false;
  }
  const lower = slug.toLowerCase();
  return lower === "champion" || lower.includes("champion-");
}

function getTopOperator(topOperators: ProfileResponse["top_operators"]): string | null {
  const all = [...topOperators.attacker, ...topOperators.defender];
  if (all.length === 0) {
    return null;
  }

  all.sort((a, b) => b.rounds_played - a.rounds_played);
  const top = all[0];

  if (!top?.operator) {
    return null;
  }
  return top.operator.charAt(0).toUpperCase() + top.operator.slice(1);
}

function getCurrentSeasonRankedStats(
  seasonModeRecords: ProfileResponse["season_mode_records"],
  currentSeason: string,
  rankedBombMode: string
): SeasonModeRecord | null {
  const seasonRecords = seasonModeRecords[currentSeason];
  if (!seasonRecords) {
    return null;
  }

  return seasonRecords[rankedBombMode] ?? null;
}

export function normaliseProfile(params: NormaliseParams): PlayerProfile {
  const { platform, response, currentSeason, rankedBombMode } = params;

  const currentSeasonRecord = response.ranked_season_records[currentSeason];

  const currentRank: RankInfo = currentSeasonRecord
    ? slugToRankInfo(
        currentSeasonRecord.rank,
        currentSeasonRecord.rank_points,
        isChampionRank(currentSeasonRecord.rank) ? currentSeasonRecord.rank_position : undefined
      )
    : slugToRankInfo(null, 0);

  const peakRankSeason: RankInfo = currentSeasonRecord
    ? slugToRankInfo(
        currentSeasonRecord.max_rank,
        currentSeasonRecord.max_rank_points,
        isChampionRank(currentSeasonRecord.max_rank)
          ? currentSeasonRecord.max_rank_position
          : undefined
      )
    : slugToRankInfo(null, 0);

  const allTimePeak = response.ranked_season_records[response.max_rank_season];
  const peakRankAllTime: RankInfo = allTimePeak
    ? slugToRankInfo(
        allTimePeak.max_rank,
        allTimePeak.max_rank_points,
        isChampionRank(allTimePeak.max_rank) ? allTimePeak.max_rank_position : undefined
      )
    : { ...peakRankSeason };

  const rankedStats = getCurrentSeasonRankedStats(
    response.season_mode_records,
    currentSeason,
    rankedBombMode
  );

  const kills = rankedStats?.kills ?? 0;
  const rawDeaths = rankedStats?.deaths ?? 0;
  const deaths = Math.max(1, rawDeaths);
  const wins = rankedStats?.wins ?? 0;
  const losses = rankedStats?.losses ?? 0;

  const leaderboardPosition =
    response.leaderboard_position !== null && response.leaderboard_position <= 10000
      ? response.leaderboard_position
      : null;

  return {
    username: response.username,
    platform,
    level: response.level,
    currentRank,
    peakRankSeason,
    peakRankAllTime,
    leaderboardPosition,
    kd: Math.round((kills / deaths) * 100) / 100,
    winRate: wins + losses > 0 ? Math.round((wins / (wins + losses)) * 100) : 0,
    kills,
    deaths: rawDeaths,
    wins,
    losses,
    headshotPercent: response.hs ?? null,
    topOperator: getTopOperator(response.top_operators),
  };
}

File: packages/r6-client/src/normalise.rs
use std::collections::BTreeMap;

use serde_json::{Map, Value};

use crate::ranks::round_js;
use crate::types::{OperatorStats, SeasonModeRecord, SeasonRankRecord};
use crate::{Platform, PlayerProfile, ProfileResponse, R6Error, RankInfo, slug_to_rank_info};

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ConfigResponse {
    pub current_season: String,
    pub ranked_bomb_mode: String,
}

fn object(value: &Value) -> Option<&Map<String, Value>> {
    value.as_object()
}

fn finite_number(value: Option<&Value>) -> Option<f64> {
    value?.as_f64().filter(|number| number.is_finite())
}

fn non_negative_number(value: Option<&Value>) -> Option<f64> {
    finite_number(value).filter(|number| *number >= 0.0)
}

fn non_negative_integer(value: Option<&Value>) -> Option<f64> {
    non_negative_number(value).filter(|number| number.fract() == 0.0)
}

fn positive_integer(value: Option<&Value>) -> Option<f64> {
    non_negative_integer(value).filter(|number| *number > 0.0)
}

fn required_nonblank<'a>(record: &'a Map<String, Value>, key: &str) -> Option<&'a str> {
    record
        .get(key)?
        .as_str()
        .filter(|value| !value.trim().is_empty())
}

pub fn parse_config_response(value: &Value) -> Result<ConfigResponse, R6Error> {
    let slugs = object(value)
        .and_then(|root| root.get("constants"))
        .and_then(object)
        .and_then(|constants| constants.get("slugs"))
        .and_then(object)
        .ok_or_else(|| R6Error::Decode("Config response is missing required slugs".into()))?;
    let current_season = required_nonblank(slugs, "current_season")
        .ok_or_else(|| R6Error::Decode("Config response is missing required slugs".into()))?;
    let ranked_bomb_mode = required_nonblank(slugs, "ranked_bomb_mode")
        .ok_or_else(|| R6Error::Decode("Config response is missing required slugs".into()))?;
    Ok(ConfigResponse {
        current_season: current_season.into(),
        ranked_bomb_mode: ranked_bomb_mode.into(),
    })
}

fn parse_rank_record(value: &Value) -> Option<SeasonRankRecord> {
    let record = object(value)?;
    let _season = record
        .get("season")?
        .as_str()
        .filter(|value| !value.is_empty())?;
    let rank = record
        .get("rank")?
        .as_str()
        .filter(|value| !value.is_empty())?;
    let max_rank = record
        .get("max_rank")?
        .as_str()
        .filter(|value| !value.is_empty())?;
    Some(SeasonRankRecord {
        rank: rank.into(),
        max_rank: max_rank.into(),
        rank_points: non_negative_number(record.get("rank_points"))?,
        max_rank_points: non_negative_number(record.get("max_rank_points"))?,
        rank_position: positive_integer(record.get("rank_position")),
        max_rank_position: positive_integer(record.get("max_rank_position")),
    })
}

fn parse_mode_record(value: &Value) -> Option<SeasonModeRecord> {
    let record = object(value)?;
    record.get("season")?.as_str()?;
    record.get("mode")?.as_str()?;
    Some(SeasonModeRecord {
        wins: non_negative_integer(record.get("wins")).unwrap_or(0.0),
        losses: non_negative_integer(record.get("losses")).unwrap_or(0.0),
        kills: non_negative_integer(record.get("kills")).unwrap_or(0.0),
        deaths: non_negative_integer(record.get("deaths")).unwrap_or(0.0),
    })
}

fn parse_operators(value: Option<&Value>) -> Vec<OperatorStats> {
    value
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
        .filter_map(|entry| {
            let record = object(entry)?;
            Some(OperatorStats {
                operator: record.get("operator")?.as_str()?.into(),
                rounds_played: non_negative_integer(record.get("rounds_played")).unwrap_or(0.0),
            })
        })
        .collect()
}

pub fn parse_profile_response(value: &Value) -> Result<ProfileResponse, R6Error> {
    let record = object(value)
        .ok_or_else(|| R6Error::Decode("Profile response is missing required fields".into()))?;
    for key in ["id", "user_id", "username", "platform", "max_rank_season"] {
        required_nonblank(record, key)
            .ok_or_else(|| R6Error::Decode("Profile response is missing required fields".into()))?;
    }
    let level = non_negative_integer(record.get("level"))
        .ok_or_else(|| R6Error::Decode("Profile response is missing required fields".into()))?;
    let rank_values = record
        .get("ranked_season_records")
        .and_then(object)
        .ok_or_else(|| R6Error::Decode("Profile response is missing required fields".into()))?;
    let mode_values = record
        .get("season_mode_records")
        .and_then(object)
        .ok_or_else(|| R6Error::Decode("Profile response is missing required fields".into()))?;
    let ranked_season_records = rank_values
        .iter()
        .filter_map(|(season, value)| {
            parse_rank_record(value).map(|parsed| (season.clone(), parsed))
        })
        .collect();
    let mut season_mode_records = BTreeMap::new();
    for (season, modes) in mode_values {
        let Some(modes) = object(modes) else { continue };
        let parsed = modes
            .iter()
            .filter_map(|(mode, value)| parse_mode_record(value).map(|entry| (mode.clone(), entry)))
            .collect();
        season_mode_records.insert(season.clone(), parsed);
    }
    let top = record.get("top_operators").and_then(object);
    Ok(ProfileResponse {
        username: record
            .get("username")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .into(),
        level,
        hs: non_negative_number(record.get("hs")).filter(|number| *number <= 100.0),
        leaderboard_position: positive_integer(record.get("leaderboard_position")),
        max_rank_season: record
            .get("max_rank_season")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .into(),
        ranked_season_records,
        season_mode_records,
        attackers: parse_operators(top.and_then(|value| value.get("attacker"))),
        defenders: parse_operators(top.and_then(|value| value.get("defender"))),
    })
}

fn is_champion(slug: &str) -> bool {
    let lower = slug.to_lowercase();
    lower == "champion" || lower.contains("champion-")
}

fn rank(record: Option<&SeasonRankRecord>, peak: bool) -> Result<RankInfo, R6Error> {
    let Some(record) = record else {
        return slug_to_rank_info(None, 0.0, None);
    };
    let (slug, rp, position) = if peak {
        (
            &record.max_rank,
            record.max_rank_points,
            record.max_rank_position,
        )
    } else {
        (&record.rank, record.rank_points, record.rank_position)
    };
    slug_to_rank_info(
        Some(slug),
        rp,
        is_champion(slug).then_some(position).flatten(),
    )
}

fn capitalize_first_utf16(value: &str) -> String {
    let mut characters = value.chars();
    let Some(first) = characters.next() else {
        return String::new();
    };
    first.to_uppercase().collect::<String>() + characters.as_str()
}

pub fn normalise_profile(
    platform: Platform,
    response: &ProfileResponse,
    current_season: &str,
    ranked_bomb_mode: &str,
) -> Result<PlayerProfile, R6Error> {
    let current_record = response.ranked_season_records.get(current_season);
    let current_rank = rank(current_record, false)?;
    let peak_rank_season = rank(current_record, true)?;
    let peak_rank_all_time = match response
        .ranked_season_records
        .get(&response.max_rank_season)
    {
        Some(record) => rank(Some(record), true)?,
        None => peak_rank_season.clone(),
    };
    let stats = response
        .season_mode_records
        .get(current_season)
        .and_then(|modes| modes.get(ranked_bomb_mode));
    let kills = stats.map_or(0.0, |value| value.kills);
    let deaths = stats.map_or(0.0, |value| value.deaths);
    let wins = stats.map_or(0.0, |value| value.wins);
    let losses = stats.map_or(0.0, |value| value.losses);
    let mut operators: Vec<&OperatorStats> = response
        .attackers
        .iter()
        .chain(&response.defenders)
        .collect();
    operators.sort_by(|left, right| right.rounds_played.total_cmp(&left.rounds_played));
    let top_operator = operators
        .first()
        .filter(|operator| !operator.operator.is_empty())
        .map(|operator| capitalize_first_utf16(&operator.operator));
    Ok(PlayerProfile {
        username: response.username.clone(),
        platform,
        level: response.level,
        current_rank,
        peak_rank_season,
        peak_rank_all_time,
        leaderboard_position: response
            .leaderboard_position
            .filter(|position| *position <= 10_000.0),
        kd: round_js(kills / deaths.max(1.0) * 100.0) / 100.0,
        win_rate: if wins + losses > 0.0 {
            round_js(wins / (wins + losses) * 100.0)
        } else {
            0.0
        },
        kills,
        deaths,
        wins,
        losses,
        headshot_percent: response.hs,
        top_operator,
    })
}
